[14.0][ADD] sale_promotion_domain_product_discount

ivs-cetmix opened the pull request.

ivs-cetmix committed:
[14.0][ADD] sale_promotion_domain_product_discount

francesco-ooops reviewed (changes requested):
when discount line is applied on website sale, each "Discount apply on" option has a different message for the discount line displayed in cart, but there is no message when the option selected is "On Domain Matching Products",

see video: https://recordit.co/JORG4kcV49

I would add the description "Discount on selected products"

francesco-ooops commented:
@isserver1 @GabbasovDinar for the rest, lgtm

francesco-ooops reviewed (approved):
Functional review ok!

elvise reviewed (approved):
LGTM!

francesco-ooops commented:
@OCA/crm-sales-marketing-maintainers review pls?

elvise commented:
@chienandalu  can you review this PR? Thanks

elvise commented:
@OCA/crm-sales-marketing-maintainers is there someone for this PR ? 😇

chienandalu commented on sale_promotion_domain_product_discount/__manifest__.py:
```diff
@@ -0,0 +1,16 @@
+    "category": "Sales Management",
+    "website": "https://github.com/OCA/sale-promotion",
+    "depends": ["sale_coupon"],
+    "external_dependencies": {},
+    "demo": [],
+    "data": [],
+    "qweb": [],
```
```suggestion
```

francesco-ooops commented:
@simahawk can you take a look at this one too? thanks :)

chienandalu commented on sale_promotion_domain_product_discount/README.rst:
```diff
@@ -0,0 +1,35 @@
+**This file is going to be generated by oca-gen-addon-readme.**
```
It's nice (not mandatory anyway) to run oca-gen-addon-readme

simahawk replied:
IMO it's easier to review the fragments, personally I always write "wait for the bot ;)" here.
In PR the docs can change, it's easier not to lose time on reviews if you don't see anything here.

chienandalu commented on sale_promotion_domain_product_discount/__manifest__.py:
```diff
@@ -0,0 +1,16 @@
+{
```
It's a common guideline to add copyright in the relevant code files. Something like:

```suggestion
# Copyright 2022 Cemtix - Ivan Sokolov
# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
{
```

chienandalu reviewed (commented):
Thanks @isserver1 for this module.

Functional review :+1: 

Some formal comments:

chienandalu commented on sale_promotion_domain_product_discount/tests/test_domain_product_discount.py:
```diff
@@ -0,0 +1,179 @@
+
+@tagged("post_install", "-at_install")
+class CetmixTestSaleCouponProgram(SavepointCase):
+    def setUp(self):
+        super(CetmixTestSaleCouponProgram, self).setUp()
```
```suggestion
    def setUpClass(cls):
        super().setUpClass()
```

simahawk replied:
and set tracking_disable for the whole suite :pray:

chienandalu commented on sale_promotion_domain_product_discount/tests/test_domain_product_discount.py:
```diff
@@ -0,0 +1,179 @@
+class CetmixTestSaleCouponProgram(SavepointCase):
+    def setUp(self):
+        super(CetmixTestSaleCouponProgram, self).setUp()
+        self.largeCabinet = self.env["product.product"].create(
```
`SavepointCase` uses the `cls` object, so you should change your case declaration like:

```suggestion
        cls.largeCabinet = cls.env["product.product"].create(
```

chienandalu commented:
Also, the guidelines recommend to name the modules with the root of the module they're extending (if any). For example, for this case, `sale_coupon_domain_product_discount` would be more appropiate

simahawk commented on sale_promotion_domain_product_discount/models/coupon_reward.py:
```diff
@@ -0,0 +1,32 @@
+from odoo import _, fields, models
```
add missing license headers everywhere

simahawk commented on sale_promotion_domain_product_discount/models/coupon_reward.py:
```diff
@@ -0,0 +1,32 @@
+    )
+
+    def name_get(self):
+        """
```
I've seen this in many places but to me it sounds weird to override the docstring of a method w/ something that is a comment explaining the override.

simahawk commented on sale_promotion_domain_product_discount/models/sale_order.py:
```diff
@@ -0,0 +1,77 @@
+        """
+        Override core method to check and apply new discount apply rule
+        """
+        if program.discount_apply_on == "domain_product":
```
please avoid such huge methods. Split the logic in smaller, more readable and more test-able methods.

ivs-cetmix replied:
I'm trying to. However this is not clear for me how to do it in this case

GabbasovDinar replied:
I have no idea how to refactor it. In any case, the original method from Odoo has more code than here.

simahawk replied:
I posted some examples... then you can use your fantasy :wink: 

Is not a blocker but your future selves and everyone else future self will be happy about this in the long term.
Is going to be more readable and easier to add unit tests and/or debug.

simahawk commented on sale_promotion_domain_product_discount/models/sale_order.py:
```diff
@@ -0,0 +1,77 @@
+                        del reward_dict[val]
+            return reward_dict.values()
+        else:
+            return super()._get_reward_values_discount(program)
```
```suggestion
        if program.discount_apply_on != "domain_product":
            return super()._get_reward_values_discount(program)
```
Do this on top and you save indentation for the whole method.

simahawk commented on sale_promotion_domain_product_discount/readme/CONFIGURE.rst:
```diff
@@ -0,0 +1 @@
+No configuration is required to use this module
```
you can trash this file then

simahawk commented on sale_promotion_domain_product_discount/readme/CREDITS.rst:
```diff
@@ -0,0 +1 @@
+* Odoo Community Association: `Icon <https://github.com/OCA/maintainer-tools/blob/master/template/module/static/description/icon.svg>`_.
```
this too

GabbasovDinar commented:
@chienandalu  Any ideas to fix in travis?
```
File "/home/travis/build/OCA/sale-promotion/sale_coupon_order_line_link/models/sale_order.py", line 73, in _link_reward_discount_lines
lines.filtered(lambda x: x.tax_id == tax).write(
UnboundLocalError: local variable 'lines' referenced before assignment
```

chienandalu commented:
> @chienandalu Any ideas to fix in travis?
> 
> ```
> File "/home/travis/build/OCA/sale-promotion/sale_coupon_order_line_link/models/sale_order.py", line 73, in _link_reward_discount_lines
> lines.filtered(lambda x: x.tax_id == tax).write(
> UnboundLocalError: local variable 'lines' referenced before assignment
> ```

Yes, as you're adding a new discount scope it doesn't enter here: https://github.com/OCA/sale-promotion/blob/14.0/sale_coupon_order_line_link/models/sale_order.py#L58-L65

I could add a hook method to get extended criterias and a default for that lines values

simahawk commented on sale_coupon_domain_product_discount/models/sale_order.py:
```diff
@@ -0,0 +1,73 @@
+        reward_dict = {}
+        amount_total = sum(self._get_base_order_lines(program).mapped("price_subtotal"))
+        for line in lines:
+            discount_line_amount = min(
+                self._get_reward_values_discount_percentage_per_line(program, line),
+                amount_total - currently_discounted_amount,
+            )
+            if discount_line_amount:
+                if line.tax_id in reward_dict:
+                    reward_dict[line.tax_id]["price_unit"] -= discount_line_amount
+                else:
+                    taxes = self.fiscal_position_id.map_tax(line.tax_id)
+                    uom_id = program.discount_line_product_id.uom_id.id
+                    reward_dict[line.tax_id] = {
+                        "name": _(
+                            "Discount: %(program)s - On product with following "
+                            "taxes: %(taxes)s",
+                            program=program.name,
+                            taxes=", ".join(taxes.mapped("name")),
+                        ),
+                        "product_id": program.discount_line_product_id.id,
+                        "price_unit": -discount_line_amount
+                        if discount_line_amount > 0
+                        else 0,
+                        "product_uom_qty": 1.0,
+                        "product_uom": uom_id,
+                        "is_reward_line": True,
+                        "tax_id": [(4, tax.id, False) for tax in taxes],
+                    }
+                    currently_discounted_amount += discount_line_amount
```
```suggestion
            self._handle_whatever_you_are_handling_here_on_a_line(program, line)
     [....]
     def _handle_whatever_you_are_handling_here_on_a_line(program, line, reward_dict, currently_discounted_amount):
            discount_line_amount = min(
                self._get_reward_values_discount_percentage_per_line(program, line),
                amount_total - currently_discounted_amount,
            )
            if discount_line_amount:
                if line.tax_id in reward_dict:
                    reward_dict[line.tax_id]["price_unit"] -= discount_line_amount
                else:
                    taxes = self.fiscal_position_id.map_tax(line.tax_id)
                    uom_id = program.discount_line_product_id.uom_id.id
                    reward_dict[line.tax_id] = {
                        "name": _(
                            "Discount: %(program)s - On product with following "
                            "taxes: %(taxes)s",
                            program=program.name,
                            taxes=", ".join(taxes.mapped("name")),
                        ),
                        "product_id": program.discount_line_product_id.id,
                        "price_unit": -discount_line_amount
                        if discount_line_amount > 0
                        else 0,
                        "product_uom_qty": 1.0,
                        "product_uom": uom_id,
                        "is_reward_line": True,
                        "tax_id": [(4, tax.id, False) for tax in taxes],
                    }
                    currently_discounted_amount += discount_line_amount
               return $whatever, you, need, to, get, back
```

simahawk commented on sale_coupon_domain_product_discount/models/sale_order.py:
```diff
@@ -0,0 +1,73 @@
+                else:
+                    taxes = self.fiscal_position_id.map_tax(line.tax_id)
+                    uom_id = program.discount_line_product_id.uom_id.id
+                    reward_dict[line.tax_id] = {
```
or move only this to `_get_reward_values_for_tax`

simahawk commented on sale_coupon_domain_product_discount/models/sale_order.py:
```diff
@@ -0,0 +1,73 @@
+        )
+        if max_amount > 0:
+            amount_already_given = 0
+            for val in list(reward_dict):
```
you could split this part to `_get_final_amount`

simahawk commented:
> > @chienandalu Any ideas to fix in travis?
> > ```
> > File "/home/travis/build/OCA/sale-promotion/sale_coupon_order_line_link/models/sale_order.py", line 73, in _link_reward_discount_lines
> > lines.filtered(lambda x: x.tax_id == tax).write(
> > UnboundLocalError: local variable 'lines' referenced before assignment
> > ```
> 
> Yes, as you're adding a new discount scope it doesn't enter here: https://github.com/OCA/sale-promotion/blob/14.0/sale_coupon_order_line_link/models/sale_order.py#L58-L65
> 
> I could add a hook method to get extended criterias and a default for that lines values

sounds like the code is just broken here: do you run tests before pushing?

chienandalu commented:
> > > @chienandalu Any ideas to fix in travis?
> > > ```
> > > File "/home/travis/build/OCA/sale-promotion/sale_coupon_order_line_link/models/sale_order.py", line 73, in _link_reward_discount_lines
> > > lines.filtered(lambda x: x.tax_id == tax).write(
> > > UnboundLocalError: local variable 'lines' referenced before assignment
> > > ```
> > 
> > 
> > Yes, as you're adding a new discount scope it doesn't enter here: https://github.com/OCA/sale-promotion/blob/14.0/sale_coupon_order_line_link/models/sale_order.py#L58-L65
> > I could add a hook method to get extended criterias and a default for that lines values
> 
> sounds like the code is just broken here: do you run tests before pushing?

Well, it's only broken in the context of a non standard discount like this one, that's why it's failing along this new module

GabbasovDinar commented:
> > @chienandalu Any ideas to fix in travis?
> > ```
> > File "/home/travis/build/OCA/sale-promotion/sale_coupon_order_line_link/models/sale_order.py", line 73, in _link_reward_discount_lines
> > lines.filtered(lambda x: x.tax_id == tax).write(
> > UnboundLocalError: local variable 'lines' referenced before assignment
> > ```
> 
> Yes, as you're adding a new discount scope it doesn't enter here: https://github.com/OCA/sale-promotion/blob/14.0/sale_coupon_order_line_link/models/sale_order.py#L58-L65
> 
> I could add a hook method to get extended criterias and a default for that lines values

Could you do that? Thanks!

elvise commented:
> > > @chienandalu Any ideas to fix in travis?
> 
> > > ```
> 
> > > File "/home/travis/build/OCA/sale-promotion/sale_coupon_order_line_link/models/sale_order.py", line 73, in _link_reward_discount_lines
> 
> > > lines.filtered(lambda x: x.tax_id == tax).write(
> 
> > > UnboundLocalError: local variable 'lines' referenced before assignment
> 
> > > ```
> 
> > 
> 
> > Yes, as you're adding a new discount scope it doesn't enter here: https://github.com/OCA/sale-promotion/blob/14.0/sale_coupon_order_line_link/models/sale_order.py#L58-L65
> 
> > 
> 
> > I could add a hook method to get extended criterias and a default for that lines values
> 
> 
> 
> Could you do that? Thanks!

@chienandalu @simahawk Can help us for this point ?

chienandalu commented:
> Can help us for this point ?

Sure, on my way :)

chienandalu commented:
> > Can help us for this point ?
> 
> Sure, on my way :)

There you are: 

- https://github.com/OCA/sale-promotion/pull/63

elvise commented:
@GabbasovDinar Merged! :-) 
can you complete this one ?

GabbasovDinar commented:
@elvise @chienandalu Ready for the review :)

elvise commented:
@simahawk how does it look to you? :)

simahawk commented on sale_coupon_domain_product_discount/models/sale_order.py:
```diff
@@ -0,0 +1,78 @@
+
+    def _get_final_amount(self, reward_dict, max_amount):
+        amount_already_given = 0
+        for val in list(reward_dict):
```
```suggestion
        _reward_dict = reward_dict.copy()
        for tax, val in reward_dict.items():
```
This way you don't alter the original dict, which sounds better for debugging.
Then in the loop alter _reward_dict instead and return it.

GabbasovDinar replied:
I'm not sure about the changes:

1. _reward_dict - pre-commit will remove the variable.
2. If we create a variable as 'reward_dict_copy' and we will use in cycle as
```python:
for tax, val in reward_dict_copy.items():
...
    if val["price_unit"] == 0:
        del reward_dict_copy[tax]
```

we will get an error: 

for example:

![image](https://user-images.githubusercontent.com/15119940/171487847-c4e50e0d-0f82-4d5f-8b44-9b7749f5d7b0.png)

or I didn't understand your idea :)

simahawk commented on sale_coupon_domain_product_discount/models/sale_order.py:
```diff
@@ -0,0 +1,78 @@
+                add_name = formatLang(
+                    self.env, max_amount, currency_obj=self.currency_id
+                )
+                reward_dict[val]["name"] += "( " + _("limited to ") + add_name + ")"
```
```suggestion
                reward_dict[val]["name"] += _("(limited to %s)", add_name)
```

simahawk reviewed (commented):
LG, few minor changes missing from my POV

GabbasovDinar commented:
@simahawk ready for review

chienandalu committed:
[PORT] sale_coupon_domain_product_discount: Port changes from 13.0 to 14.0

TT36131

[IMP] sale_coupon_domain_product_discount: strict limits

[FIX] fix lint errors

elvise commented:
@simahawk looks good now ? :)

elvise commented:
@simahawk sorry if ping you again, but pending only your review :)

GabbasovDinar commented:
@chienandalu I ported your changes from 13.0

chienandalu reviewed (approved):
Great! Good team up! :smile:

pedrobaeza commented:
I think you can squash last 3 commits into one. It sounds weird to introduce a diff with 13.0.1.0.0, and again another to 14.0.1.0.0

francesco-ooops reviewed (approved):
Functional review ok!

francesco-ooops commented:
@simahawk @pedrobaeza this PR is now the porting of #64 , would you say it's ready for merge?

simahawk commented on sale_coupon_domain_product_discount/views/coupon_program_views.xml:
```diff
@@ -0,0 +1,30 @@
+            >
+                <attribute
+                    name="attrs"
+                >{'invisible': ['|', '|', '|', ('discount_apply_on', '!=', 'specific_products'),('discount_type', '!=', 'percentage'), ('reward_type', 'in', ('product', 'free_shipping')), ('discount_apply_on_domain_product', '!=', False)], 'required': [('reward_type', '=', 'discount'),('discount_apply_on', '=', 'specific_products'),('discount_type', '=', 'percentage'),('discount_apply_on_domain_p ...
```
next time you can define a computed boolean field and use only this to show/hide

simahawk commented:
/ocabot merge minor
